redisson/src/main/java/org/redisson/RedissonPermitExpirableSemaphore.java

```diff
@@ -645,11 +655,79 @@ public RFuture<Integer> availablePermitsAsync() {
                 Arrays.<Object>asList(getRawName(), timeoutName, getChannelName()), System.currentTimeMillis());
     }
 
+    @Override
+    public RFuture<Integer> getPermitsAsync() {
+        return commandExecutor.evalWriteAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_INTEGER,
+                "local expiredIds = redis.call('zrangebyscore', KEYS[2], 0, ARGV[1], 'limit', 0, -1); " +
+                "if #expiredIds > 0 then " +
+                    "redis.call('zrem', KEYS[2], unpack(expiredIds)); " +
+                    "local value = redis.call('incrby', KEYS[1], #expiredIds); " +
+                    "if tonumber(value) > 0 then " +
+                        "redis.call('publish', KEYS[3], value); " +
+                    "end;" +
+                "end; " +
+                "local available = redis.call('get', KEYS[1]); " +
+                "local claimed = redis.call('zcount', KEYS[2], 0, '+inf'); " +
+                "if available == false then " +
+                    "return 0 " +
+                "end;" +
+                "if claimed == false then " +
+                    "return tonumber(available) " +
+                "end;" +
+                "return tonumber(available) + claimed;",
+                Arrays.<Object>asList(getRawName(), timeoutName, getChannelName()), System.currentTimeMillis());
+    }
+
+    @Override
+    public RFuture<Integer> claimedPermitsAsync() {
+        return commandExecutor.evalWriteAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_INTEGER,
+                "local expiredIds = redis.call('zrangebyscore', KEYS[2], 0, ARGV[1], 'limit', 0, -1); " +
+                "if #expiredIds > 0 then " +
+                    "redis.call('zrem', KEYS[2], unpack(expiredIds)); " +
+                    "local value = redis.call('incrby', KEYS[1], #expiredIds); " +
+                    "if tonumber(value) > 0 then " +
+                        "redis.call('publish', KEYS[3], value); " +
+                    "end;" +
+                "end; " +
+                "local claimed = redis.call('zcount', KEYS[2], 0, '+inf'); " +
+                "return claimed == false and 0 or claimed;",
+                Arrays.<Object>asList(getRawName(), timeoutName, getChannelName()), System.currentTimeMillis());
+    }
+
     @Override
     public boolean trySetPermits(int permits) {
         return get(trySetPermitsAsync(permits));
     }
 
+    @Override
+    public int setPermits(int permits) {
+        return get(setPermitsAsync(permits));
+    }
+
+    @Override
+    public RFuture<Integer> setPermitsAsync(int permits) {
+        return commandExecutor.evalWriteAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_INTEGER,
+                "local available = redis.call('get', KEYS[1]); " +
```

Comment:
It's expected that `setPermits` just override the permits value. Like `setMaxSize` and `trySetMaxSize` methods work for `RMapCache` object

Reply:
I'm not sure if I follow your suggestion. 

In order to match the semantics of RMapCache's setMaxSize, I believe we do need to take into consideration the number of claimed permits, like the implementation in the PR does.

Unlike the max size of the `RMapCache` object, the max number of permits in `RPermitExpirableSemaphore`  is not isolated and stored in a separate entry. After the max number of permits is set the very first time, that value becomes the number of available permits and can changed with each `acquire`/`release`.

Because of this, when we re-set the max using `setPermits`, we need to consider the number of claimed permits, too, or we could miscalculate the total number of permits.

---
I’ve included an example to walk through the difference between the behaviours. This is also captured in [the test cases](https://github.com/redisson/redisson/pull/4719/files#diff-8d51b6c35753a38648e0759011ae9ebb83d954359ba65801f5cd117a647ea5f9R216-R225).

Suppose we create a semaphore which initially has a maximum of 10 permits.

```
RPermitExpirableSemaphore s = redisson.getPermitExpirableSemaphore(“example”);
s.trySetPermits(10); 

s.getPermits(); // 10
s.availablePermits(); // 10
s.claimedPermits(); // 0
```

Next the client acquires two permits. This decrements [the value stored for the number of available permits](https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/RedissonPermitExpirableSemaphore.java#L362).

```
String permit1 = s.acquire();
String permit2 = s.acquire();

s.getPermits(); // 10
s.availablePermits(); // 8
s.claimedPermits(); // 2
```

Suppose we now want to change the max number of permits to 20. To me, this means that the max number of permits which can be acquired, in aggregate, across all clients is 20. Including the two which are currently acquired by clients.

If we implement `setPermits(int permit)` such that it only directly overrides the value used as the number of available permits with the max value passed to `setPermits`, we would have updated the values as seen below:
```
s.setPermits(20);

s.getPermits(); // 22
s.availablePermits(); // 20
s.claimedPermits(); // 2
``` 

Then, when our client release the two permits it was holding, it will [increment the number of available permits](https://github.com/redisson/redisson/blob/master/redisson/src/main/java/org/redisson/RedissonPermitExpirableSemaphore.java#L566) and we'd see the following:

```
s.release(permit1);
s.release(permit2);

s.getPermits(); // 22
s.availablePermits(); // 22
s.claimedPermits(); // 0
```

--- 

What is implemented in the PR currently for `setPermits` takes the number of claimed permits into consideration such that we have the following, which I believe is what we want.

```
s.setPermits(20);

s.getPermits(); // 20
s.availablePermits(); // 18
s.claimedPermits(); // 2

s.release(permit1);
s.release(permit2);

s.getPermits(); // 20
s.availablePermits(); // 20
s.claimedPermits(); // 0
```

Let me know if this is not what you had in mind - we may need to clarify our definitions of max permits vs. total permits vs. available permits.